text/0000-freeze-uninit.md

```diff
@@ -0,0 +1,217 @@
+    * If intermediate representations are cooperative, it may be beneficial to provide the operation in the future, as it could perform only the writes required to ensure the backing memory is in a stable state (such as 1 write every 4096 bytes)
+* `MaybeUninit<Int>` (and maybe `MaybeUninit<Float>`) could have arithmetic that is defined as `uninit (op) x` or `x (op) uninit` is `uninit`.
+    * While this can partially solve the second use case (by following it through `Simd`) and the third use case, this does not help the first use case
+    * This RFC does not preclude these operations from being defined in the future, and may even make them more useful.
+* `MaybeUninit::freeze` could instead be `pub unsafe fn freeze(self) -> T`, like `assume_init`.
+    * While this would allow `mu.freeze().assume_init()` to be written in fewer lines of code, maintaining the value as `MaybeUninit` may make some code possible using careful offsetting and `MaybeUninit::as_ptr()`/`MaybeUninit::as_mut_ptr()`.
+    * Most uses of `MaybeUninit` would require immediately using `.assume_init()` on the result, however, this is a potential footgun if `T` has any invalid initialized values, or is a user-defined type with a complex safety invariant. It is hoped that the extra verbosity, on top of helpful documentation, will allow intermediate, advanced, or expert unsafe code users making use of `MaybeUninit::freeze` to recognize this potential footgun and to carefully validate the types involved. 
+
+# Prior art
+[prior-art]: #prior-art
+
+[LLVM](https://llvm.org/docs/LangRef.html#freeze-instruction) supports freeze by-value. 
+
+
+* GCC may support an equivalent operation via the `SAVE_EXPR` GIMPLE code.
+* See <https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem/topic/GCC.20and.20freeze>
+
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+* Which of the library functions should recieve the direct language intrinsic, between `ptr::read_freeze` and `MaybeUninit::freeze`
+* Should the `ptr::read_freeze` and `MaybeUninit::freeze` functions be `const`
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+* With the project-portable-simd, the `Simd` type could support `Simd::load_partial<T, const N: usize>(x: *const T) -> Simd<[MaybeUninit<T>;N]>` (signature to be bikeshed) which could then be frozen lanewise into `Simd<[T;N]>`. With proper design (which is not the subject of this RFC), this could allow optimized loads at allocation boundaries by allowing operations that may physically perform an out-of-bounds read, but instead logically returns uninit for the out-of-bounds portion. This can be used to write an optimized implementation of `memchr`, `strchr`, or `strlen`, or even optimize `UTF-8` encoding and processing.
```

Comment:
This relies on a hypothetical new LLVM operation that does a load where out-of-bounds memory is `undef`, right? That should be mentioned.

Reply:
Or an appropriate masked/partial load that doesn't logically perform the problematic reads at the llvm level.

Reply:
The text does not indicate a way for portable-simd to have the information to produce such a mask.

Reply:
Suggestion: use an expository signature of

```rust
Simd::read_select(
    source: *const [T],
    enable: Mask<isize, N>,
) -> Simd<MaybeUninit<T>, N>
```

This is deliberately imitating the signature of [`gather_select`](https://doc.rust-lang.org/stable/std/simd/prelude/struct.Simd.html#method.gather_select) and [`gather_select_ptr`](https://doc.rust-lang.org/stable/std/simd/prelude/struct.Simd.html#method.gather_select_ptr). While likely not the "correct" signature for a partial read, it can clearly do what the example usage wants to do.

Reply:
`load_select` is likely closer to the intended primitive.